### ISSUE TITLE
Can't configure functions in options maps on command line or in files.

### ISSUE THREAD

**Issue by kkinnear** [kkinnear/zprint#124]
There are several places where functions are allowed in an options map.  If you configure the functions in the options map using `set-options!`, that works fine.  But that only works if you are using zprint as a library.  If you use the zprint as an uberjar or a pre-built graalVM binary, then the only way you can configure it is with `.zprintrc` files or from the command line.  In either case, functions (or any executable code snippets) are not interpreted correctly. 

The challenge to "fixing" this is that allowing executable code in an options map loaded from a file opens the door to a number of unpleasant security exploits.  How to balance using functions to enrich and adapt the capabilities of zprint with the potential security issues raised by executable code loaded from files is, as yet, undetermined.

I can imagine requiring an explicit marker on the command line to allow executable code to be loaded from a `.zprintrc` file or the command line options map.  But would that still be too risky for people?

I'm very interested in suggestions and discussion on this topic.

**Comment by agriffis** [comment]
There are four sources of configuration, and we only need to worry about two of them:

1. `~/.zprintrc` is **trusted** configuration, because it is outside the untrusted code, and writing to it implies that security is broken by another vector for which zprint is not responsible.

2. zprint command-line is **trusted** configuration, because it is part of the execution context, and anything that can execute commands is already able to run what it wants. If there are malicious command-line options, once again that implies that security is broken by another vector for which zprint is not responsible.

3. `.zprintrc` in a repository is **untrusted** configuration by default

4. `;!zprint` comment is **untrusted** configuration by default

The concern here is about running *untrusted* code *unexpectedly*. We bring untrusted code onto our machines all the time, cloning GitHub repos for example, but normally we expect that's okay because the code won't run directly. In fact, that's pretty much an illusion. We almost immediately proceed to running `./configure` or `make` or `npm install` or what-have-you. Most of these will be opportunities for a malevolent or compromised repository to execute arbitrary code.

However zprint presents a unique opportunity for malicious code to execute. If we clone a repo and then want to inspect the code before running it, perhaps we'd like to reformat it for readability, or perhaps we have our editor configured to use zprint for formatting. In either case, we don't expect malicious code to run.

The situation for zprint isn't really unique, it's effectively the same as for editors, and they've been dealing with this for a while. Vim trusts command-line and `~/.vimrc` by default, doesn't support repository-local configuration at all, and restricts what can be configured via in-file modelines. A number of years ago there were security issues for modelines. Those were solved by two approaches: (1) restrict what can be done in modelines, (2) disallow modelines unless trusted configuration has `set modeline`

Emacs trusts command-line and `~/.emacs` by default, and only supports "safe" settings in `.dir-locals.el` and file variables. For the latter, if it encounters settings it deems unsafe, it asks the user before accepting them (or assumes no if there's no opportunity to ask).

Since it's always safe for zprint to *read* configuration, I think that zprint can behave more like Emacs than like Vim. I wish that `:search-config? true` were the default instead of needing to be configured, and perhaps it's not too late to change that. However zprint needs a concept of trusting repo-local configuration for functions, including `.zprintrc` or the comment API. That trust needs to be configured in trusted config, meaning `~/.zprintrc` or command-line.

My inclination would be to have a list of path globs, something like this in `.zprintrc`

```clojure
{:map {:comma? false}
 :search-config? true
 :trusted-configs ["/home/aron/work/**" "/home/aron/foo/.zprintrc"]}
```

With this approach, you can check if a given source of configuration, whether comment API or a found `.zprintrc`, matches one of the globs. If it doesn't match, then ignore functions from that configuration source. (Maybe emit a warning? Not sure about that part.)

Just make sure to concat `:trusted-configs` from the command-line as well. Editors need to be able to supply trusted configuration to plugins without needing to write `~/.zprintrc`, so that the concept of a trusted path doesn't need to be configured by the user separately for every code formatter and helper tool. At that point it becomes the editor's responsibility to track trusted paths and convey that to plugins.

**Comment by kkinnear** [comment]
Wow, thanks for the deep analysis here.  I'm still digesting this and how I would implement it, but you make some compelling points. Thanks again for bringing a larger perspective than I happen to have to the discussion!

**Comment by kkinnear** [comment]
Fixed this as much as possible, though it is still rather messed up in the graalVM binaries.  For now fixed as much as it is going to get in 0.5.4.
